pkg/parsers/operatingsystem/operatingsystem_linux_test.go

```diff
@@ -38,10 +43,7 @@ PRETTY_NAME=Ubuntu 14.04.LTS`,
 		},
 	}
 
-	valids := []struct {
```

Comment:
Instead of having `valid` and `invalid` types, perhaps just combine them in a single list. I gave this a quick spin (also turning them into subtests), but let me know what you think;

```patch
diff --git a/pkg/parsers/operatingsystem/operatingsystem_linux_test.go b/pkg/parsers/operatingsystem/operatingsystem_linux_test.go
index 2eb9d47673..5d35cd4024 100644
--- a/pkg/parsers/operatingsystem/operatingsystem_linux_test.go
+++ b/pkg/parsers/operatingsystem/operatingsystem_linux_test.go
@@ -7,45 +7,39 @@ import (
 	"os"
 	"path/filepath"
 	"testing"
-)
-
-type valid struct {
-	content  string
-	expected string
-}
 
-type invalid struct {
-	content       string
-	expectedError string
-}
+	"gotest.tools/assert"
+)
 
 func TestGetOperatingSystem(t *testing.T) {
-	invalids := []invalid{
+	tests := []struct {
+		name        string
+		content     string
+		expected    string
+		expectedErr string
+	}{
 		{
-			`PRETTY_NAME=Source Mage GNU/Linux
+			content: `PRETTY_NAME=Source Mage GNU/Linux
 PRETTY_NAME=Ubuntu 14.04.LTS`,
-			"PRETTY_NAME needs to be enclosed by quotes if they have spaces: Source Mage GNU/Linux",
+			expectedErr: "PRETTY_NAME needs to be enclosed by quotes if they have spaces: Source Mage GNU/Linux",
 		},
 		{
-			`PRETTY_NAME="Ubuntu Linux
+			content: `PRETTY_NAME="Ubuntu Linux
 PRETTY_NAME=Ubuntu 14.04.LTS`,
-			"PRETTY_NAME is invalid: invalid command line string",
+			expectedErr: "PRETTY_NAME is invalid: invalid command line string",
 		},
 		{
-			`PRETTY_NAME=Ubuntu'
+			content: `PRETTY_NAME=Ubuntu'
 PRETTY_NAME=Ubuntu 14.04.LTS`,
-			"PRETTY_NAME is invalid: invalid command line string",
+			expectedErr: "PRETTY_NAME is invalid: invalid command line string",
 		},
 		{
-			`PRETTY_NAME'
+			content: `PRETTY_NAME'
 PRETTY_NAME=Ubuntu 14.04.LTS`,
-			"PRETTY_NAME needs to be enclosed by quotes if they have spaces: Ubuntu 14.04.LTS",
+			expectedErr: "PRETTY_NAME needs to be enclosed by quotes if they have spaces: Ubuntu 14.04.LTS",
 		},
-	}
-
-	valids := []valid{
 		{
-			`NAME="Ubuntu"
+			content: `NAME="Ubuntu"
 PRETTY_NAME_AGAIN="Ubuntu 14.04.LTS"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
@@ -54,10 +48,10 @@ VERSION_ID="14.04"
 HOME_URL="http://www.ubuntu.com/"
 SUPPORT_URL="http://help.ubuntu.com/"
 BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"`,
-			"Linux",
+			expected: "Linux",
 		},
 		{
-			`NAME="Ubuntu"
+			content: `NAME="Ubuntu"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
 ID_LIKE=debian
@@ -65,10 +59,10 @@ VERSION_ID="14.04"
 HOME_URL="http://www.ubuntu.com/"
 SUPPORT_URL="http://help.ubuntu.com/"
 BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"`,
-			"Linux",
+			expected: "Linux",
 		},
 		{
-			`NAME=Gentoo
+			content: `NAME=Gentoo
 ID=gentoo
 PRETTY_NAME="Gentoo/Linux"
 ANSI_COLOR="1;32"
@@ -76,10 +70,10 @@ HOME_URL="http://www.gentoo.org/"
 SUPPORT_URL="http://www.gentoo.org/main/en/support.xml"
 BUG_REPORT_URL="https://bugs.gentoo.org/"
 `,
-			"Gentoo/Linux",
+			expected: "Gentoo/Linux",
 		},
 		{
-			`NAME="Ubuntu"
+			content: `NAME="Ubuntu"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
 ID_LIKE=debian
@@ -88,48 +82,62 @@ VERSION_ID="14.04"
 HOME_URL="http://www.ubuntu.com/"
 SUPPORT_URL="http://help.ubuntu.com/"
 BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"`,
-			"Ubuntu 14.04 LTS",
+			expected: "Ubuntu 14.04 LTS",
 		},
 		{
-			`NAME="Ubuntu"
+			content: `NAME="Ubuntu"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
 ID_LIKE=debian
 PRETTY_NAME='Ubuntu 14.04 LTS'`,
-			"Ubuntu 14.04 LTS",
+			expected: "Ubuntu 14.04 LTS",
 		},
 		{
-			`PRETTY_NAME=Source
+			content: `PRETTY_NAME=Source
 NAME="Source Mage"`,
-			"Source",
+			expected: "Source",
 		},
 		{
-			`PRETTY_NAME=Source
+			content: `PRETTY_NAME=Source
 PRETTY_NAME="Source Mage"`,
-			"Source Mage",
+			expected: "Source Mage",
 		},
 	}
 
-	testParsingOsRelease(t, GetOperatingSystem, invalids, valids)
+	defer setupTest()()
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if err := ioutil.WriteFile(etcOsRelease, []byte(test.content), 0600); err != nil {
+				t.Fatalf("failed to write to %s: %v", etcOsRelease, err)
+			}
+			s, err := GetOperatingSystem()
+			if test.expectedErr == "" {
+				assert.NilError(t, err)
+			} else {
+				assert.Error(t, err, test.expectedErr)
+			}
+			assert.Equal(t, s, test.expected)
+		})
+	}
 }
 
 func TestGetOperatingSystemVersion(t *testing.T) {
-	invalids := []invalid{
+	tests := []struct {
+		name        string
+		content     string
+		expected    string
+		expectedErr string
+	}{
 		{
-			`VERSION_ID="18.04
+			name: "invalid version id",
+			content: `VERSION_ID="18.04
 VERSION_ID=18.04`,
-			"VERSION_ID is invalid: invalid command line string",
+			expectedErr: "VERSION_ID is invalid: invalid command line string",
 		},
 		{
-			`VERSION_ID=18.04'
-VERSION_ID=18.04`,
-			"VERSION_ID is invalid: invalid command line string",
-		},
-	}
-
-	valids := []valid{
-		{
-			`NAME="Ubuntu"
+			name: "ubuntu 14.04",
+			content: `NAME="Ubuntu"
 PRETTY_NAME="Ubuntu 14.04.LTS"
 VERSION="14.04, Trusty Tahr"
 ID=ubuntu
@@ -138,10 +146,11 @@ VERSION_ID="14.04"
 HOME_URL="http://www.ubuntu.com/"
 SUPPORT_URL="http://help.ubuntu.com/"
 BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"`,
-			"14.04",
+			expected: "14.04",
 		},
 		{
-			`NAME=Gentoo
+			name: "gentoo",
+			content: `NAME=Gentoo
 ID=gentoo
 PRETTY_NAME="Gentoo/Linux"
 ANSI_COLOR="1;32"
@@ -149,47 +158,42 @@ HOME_URL="http://www.gentoo.org/"
 SUPPORT_URL="http://www.gentoo.org/main/en/support.xml"
 BUG_REPORT_URL="https://bugs.gentoo.org/"
 `,
-			"",
 		},
 		{
-			`VERSION_ID="14.04"
+			name: "dual version id",
+			content: `VERSION_ID="14.04"
 VERSION_ID=18.04`,
-			"18.04",
+			expected: "18.04",
 		},
 	}
 
-	testParsingOsRelease(t, GetOperatingSystemVersion, invalids, valids)
+	defer setupTest()()
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if err := ioutil.WriteFile(etcOsRelease, []byte(test.content), 0600); err != nil {
+				t.Fatalf("failed to write to %s: %v", etcOsRelease, err)
+			}
+			s, err := GetOperatingSystemVersion()
+			if test.expectedErr == "" {
+				assert.NilError(t, err)
+			} else {
+				assert.Error(t, err, test.expectedErr)
+			}
+			assert.Equal(t, s, test.expected)
+		})
+	}
 }
 
-func testParsingOsRelease(t *testing.T, funcToTest func() (string, error), invalids []invalid, valids []valid) {
+func setupTest() func() {
 	var backup = etcOsRelease
 
 	dir := os.TempDir()
 	etcOsRelease = filepath.Join(dir, "etcOsRelease")
 
-	defer func() {
+	return func() {
 		os.Remove(etcOsRelease)
 		etcOsRelease = backup
-	}()
-
-	for _, elt := range invalids {
-		if err := ioutil.WriteFile(etcOsRelease, []byte(elt.content), 0600); err != nil {
-			t.Fatalf("failed to write to %s: %v", etcOsRelease, err)
-		}
-		s, err := funcToTest()
-		if err == nil || err.Error() != elt.expectedError {
-			t.Errorf("Expected an error %q, got %q (err: %v)", elt.expectedError, s, err)
-		}
-	}
-
-	for _, elt := range valids {
-		if err := ioutil.WriteFile(etcOsRelease, []byte(elt.content), 0600); err != nil {
-			t.Fatalf("failed to write to %s: %v", etcOsRelease, err)
-		}
-		s, err := funcToTest()
-		if err != nil || s != elt.expected {
-			t.Errorf("Expected %q, got %q (err: %v)", elt.expected, s, err)
-		}
 	}
 }
 ```

Reply:
Thanks, I like it. DRY-ed up just a bit further.